File: frontend/src/services/propertyIntelligence.ts
import apiService from "./api";
import {
  BulkPropertyAnalysisRequest,
  PropertyAnalyticsRequest,
  PropertyAnalyticsResponse,
  PropertyComparisonResult,
  PropertyMarketInsight,
  PropertyProfile,
  PropertySearchRequest,
  PropertySearchResponse,
  PropertyWatchlistItem,
  AlertPreferences
} from "../types";

export class PropertyIntelligenceService {
  private static instance: PropertyIntelligenceService;
  private baseUrl = "/api/property-intelligence";

  public static getInstance(): PropertyIntelligenceService {
    if (!PropertyIntelligenceService.instance) {
      PropertyIntelligenceService.instance = new PropertyIntelligenceService();
    }
    return PropertyIntelligenceService.instance;
  }

  /**
   * Advanced property search with intelligent filtering
   */
  async searchProperties(
    request: PropertySearchRequest,
  ): Promise<PropertySearchResponse> {
    try {
      const response = await apiService.post(`${this.baseUrl}/search`, request);
      return response.data;
    } catch (error) {
      console.error("Property search failed:", error);
      throw new Error("Failed to search properties");
    }
  }

  /**
   * Comprehensive property analysis
   */
  async analyzeProperty(
    request: PropertyAnalyticsRequest,
  ): Promise<PropertyAnalyticsResponse> {
    try {
      const response = await apiService.post(
        `${this.baseUrl}/analyze`,
        request,
      );
      return response.data;
    } catch (error) {
      console.error("Property analysis failed:", error);
      throw new Error("Failed to analyze property");
    }
  }

  /**
   * Bulk property portfolio analysis
   */
  async bulkAnalyzeProperties(
    request: BulkPropertyAnalysisRequest,
  ): Promise<unknown> {
    try {
      const response = await apiService.post(
        `${this.baseUrl}/bulk-analyze`,
        request,
      );
      return response.data;
    } catch (error) {
      console.error("Bulk analysis failed:", error);
      throw new Error("Failed to perform bulk analysis");
    }
  }

  /**
   * Get user's property watchlist
   */
  async getWatchlist(
    limit: number = 50,
    offset: number = 0,
  ): Promise<PropertyWatchlistItem[]> {
    try {
      const response = await apiService.get(`${this.baseUrl}/watchlist`, {
        params: { limit, offset },
      });
      return response.data;
    } catch (error) {
      console.error("Failed to get watchlist:", error);
      throw new Error("Failed to retrieve property watchlist");
    }
  }

  /**
   * Add property to watchlist
   */
  async addToWatchlist(property: {
    address: string;
    notes?: string;
    tags?: string[];
    alert_preferences?: AlertPreferences;
  }): Promise<unknown> {
    try {
      const response = await apiService.post(
        `${this.baseUrl}/watchlist`,
        property,
      );
      return response.data;
    } catch (error) {
      console.error("Failed to add to watchlist:", error);
      throw new Error("Failed to add property to watchlist");
    }
  }

  /**
   * Remove property from watchlist
   */
  async removeFromWatchlist(watchlistItemId: string): Promise<void> {
    try {
      await apiService.delete(`${this.baseUrl}/watchlist/${watchlistItemId}`);
    } catch (error) {
      console.error("Failed to remove from watchlist:", error);
      throw new Error("Failed to remove property from watchlist");
    }
  }

  /**
   * Get market insights for location
   */
  async getMarketInsights(
    location: string,
    insightTypes: string[] = ["trends", "forecasts"],
    limit: number = 10,
  ): Promise<PropertyMarketInsight[]> {
    try {
      const response = await apiService.get(`${this.baseUrl}/market-insights`, {
        params: { location, insight_types: insightTypes.join(","), limit },
      });
      return response.data;
    } catch (error) {
      console.error("Failed to get market insights:", error);
      throw new Error("Failed to retrieve market insights");
    }
  }

  /**
   * Compare multiple properties
   */
  async compareProperties(
    properties: string[],
    comparisonCriteria: string[] = ["price", "investment", "growth"],
  ): Promise<PropertyComparisonResult> {
    try {
      const response = await apiService.post(`${this.baseUrl}/compare`, {
        properties,
        comparison_criteria: comparisonCriteria,
      });
      return response.data;
    } catch (error) {
      console.error("Property comparison failed:", error);
      throw new Error("Failed to compare properties");
    }
  }

  /**
   * Export property data in various formats
   */
  async exportPropertyData(
    format: "csv" | "json" | "pdf",
    properties: string[],
  ): Promise<Blob> {
    try {
      const response = await apiService.post(`${this.baseUrl}/export`, {
        format,
        properties,
      }, {
        responseType: "blob",
      });
      return response.data;
    } catch (error) {
      console.error("Export failed:", error);
      throw new Error("Failed to export property data");
    }
  }

  /**
   * Get comprehensive property analysis
   */
  async getPropertyAnalysis(propertyId: string): Promise<PropertyProfile> {
    try {
      const response = await apiService.get(
        `${this.baseUrl}/property/${propertyId}/analysis`,
      );
      return response.data;
    } catch (error) {
      console.error("Failed to get property analysis:", error);
      throw new Error("Failed to retrieve property analysis");
    }
  }

  /**
   * Get market analysis for suburb
   */
  async getMarketAnalysis(
    suburb: string,
    state: string,
    propertyType?: string,
  ): Promise<unknown> {
    try {
      const response = await apiService.get(`${this.baseUrl}/market-analysis`, {
        params: { suburb, state, property_type: propertyType },
      });
      return response.data;
    } catch (error) {
      console.error("Failed to get market analysis:", error);
      throw new Error("Failed to retrieve market analysis");
    }
  }

  /**
   * Get investment opportunities
   */
  async getInvestmentOpportunities(
    location: string,
    budgetRange?: [number, number],
    investmentStrategy: string = "growth",
  ): Promise<unknown> {
    try {
      const response = await apiService.get(
        `${this.baseUrl}/investment-opportunities`,
        {
          params: {
            location,
            min_budget: budgetRange?.[0],
            max_budget: budgetRange?.[1],
            strategy: investmentStrategy,
          },
        },
      );
      return response.data;
    } catch (error) {
      console.error("Failed to get investment opportunities:", error);
      throw new Error("Failed to retrieve investment opportunities");
    }
  }

  /**
   * Get property valuation
   */
  async getPropertyValuation(
    address: string,
    valuationType: "avm" | "desktop" | "professional" = "avm",
  ): Promise<unknown> {
    try {
      const response = await apiService.post(`${this.baseUrl}/valuation`, {
        address,
        valuation_type: valuationType,
      });
      return response.data;
    } catch (error) {
      console.error("Failed to get property valuation:", error);
      throw new Error("Failed to retrieve property valuation");
    }
  }

  /**
   * Get suburb statistics
   */
  async getSuburbStatistics(suburb: string, state: string): Promise<unknown> {
    try {
      const response = await apiService.get(
        `${this.baseUrl}/suburb-statistics`,
        {
          params: { suburb, state },
        },
      );
      return response.data;
    } catch (error) {
      console.error("Failed to get suburb statistics:", error);
      throw new Error("Failed to retrieve suburb statistics");
    }
  }

  /**
   * Advanced property search with filters
   */
  async searchPropertiesAdvanced(filters: {
    location?: string;
    minPrice?: number;
    maxPrice?: number;
    propertyTypes?: string[];
    minBedrooms?: number;
    maxBedrooms?: number;
    features?: string[];
    sortBy?: string;
    limit?: number;
  }): Promise<PropertySearchResponse> {
    try {
      const searchRequest: PropertySearchRequest = {
        query: filters.location,
        filters: {
          min_price: filters.minPrice,
          max_price: filters.maxPrice,
          property_types: filters.propertyTypes || [],
          min_bedrooms: filters.minBedrooms,
          max_bedrooms: filters.maxBedrooms,
          features_required: filters.features || [],
          suburbs: filters.location ? [filters.location] : [],
          states: [],
          min_carspaces: undefined,
          min_land_area: undefined,
          max_land_area: undefined,
        },
        location: filters.location,
        radius_km: 50,
        limit: filters.limit || 20,
        sort_by: (filters.sortBy as PropertySearchRequest['sort_by']) || "relevance",
        include_off_market: false,
        include_historical: false,
      };

      return await this.searchProperties(searchRequest);
    } catch (error) {
      console.error("Advanced property search failed:", error);
      throw new Error("Failed to search properties with advanced filters");
    }
  }

  /**
   * Get property price history
   */
  async getPropertyPriceHistory(propertyId: string): Promise<unknown> {
    try {
      const response = await apiService.get(
        `${this.baseUrl}/property/${propertyId}/price-history`,
      );
      return response.data;
    } catch (error) {
      console.error("Failed to get price history:", error);
      throw new Error("Failed to retrieve property price history");
    }
  }

  /**
   * Get rental estimates for property
   */
  async getRentalEstimate(address: string): Promise<unknown> {
    try {
      const response = await apiService.post(
        `${this.baseUrl}/rental-estimate`,
        { address },
      );
      return response.data;
    } catch (error) {
      console.error("Failed to get rental estimate:", error);
      throw new Error("Failed to retrieve rental estimate");
    }
  }

  /**
   * Get property report (PDF)
   */
  async generatePropertyReport(
    propertyId: string,
    reportType: "basic" | "standard" | "premium" = "standard",
  ): Promise<Blob> {
    try {
      const response = await apiService.post(
        `${this.baseUrl}/property/${propertyId}/report`,
        {
          report_type: reportType,
        },
        {
          responseType: "blob",
        },
      );
      return response.data;
    } catch (error) {
      console.error("Failed to generate property report:", error);
      throw new Error("Failed to generate property report");
    }
  }

  /**
   * Get nearby amenities for property
   */
  async getNearbyAmenities(
    address: string,
    radius: number = 2000,
  ): Promise<unknown> {
    try {
      const response = await apiService.get(
        `${this.baseUrl}/nearby-amenities`,
        {
          params: { address, radius },
        },
      );
      return response.data;
    } catch (error) {
      console.error("Failed to get nearby amenities:", error);
      throw new Error("Failed to retrieve nearby amenities");
    }
  }

  /**
   * Get school catchments for property
   */
  async getSchoolCatchments(address: string): Promise<unknown> {
    try {
      const response = await apiService.get(
        `${this.baseUrl}/school-catchments`,
        {
          params: { address },
        },
      );
      return response.data;
    } catch (error) {
      console.error("Failed to get school catchments:", error);
      throw new Error("Failed to retrieve school catchments");
    }
  }

  /**
   * Get property ownership history
   */
  async getOwnershipHistory(propertyId: string): Promise<unknown> {
    try {
      const response = await apiService.get(
        `${this.baseUrl}/property/${propertyId}/ownership-history`,
      );
      return response.data;
    } catch (error) {
      console.error("Failed to get ownership history:", error);
      throw new Error("Failed to retrieve property ownership history");
    }
  }

  /**
   * Get market trends data
   */
  async getMarketTrends(
    timeframe: "3M" | "6M" | "1Y" | "2Y" = "1Y",
  ): Promise<unknown> {
    try {
      const response = await apiService.get(`${this.baseUrl}/market-trends`, {
        params: { timeframe },
      });
      return response.data;
    } catch (error) {
      console.error("Failed to get market trends:", error);
      throw new Error("Failed to retrieve market trends");
    }
  }

  /**
   * Get national market statistics
   */
  async getNationalMarketStats(): Promise<unknown> {
    try {
      const response = await apiService.get(`${this.baseUrl}/market/national`);
      return response.data;
    } catch (error) {
      console.error("Failed to get national market stats:", error);
      throw new Error("Failed to retrieve national market statistics");
    }
  }

  /**
   * Get state market comparison data
   */
  async getStateMarketComparison(): Promise<unknown> {
    try {
      const response = await apiService.get(`${this.baseUrl}/market/states`);
      return response.data;
    } catch (error) {
      console.error("Failed to get state market comparison:", error);
      throw new Error("Failed to retrieve state market comparison");
    }
  }

  /**
   * Get high growth suburbs
   */
  async getHotSuburbs(limit: number = 10): Promise<unknown> {
    try {
      const response = await apiService.get(
        `${this.baseUrl}/market/hot-suburbs`,
        {
          params: { limit },
        },
      );
      return response.data;
    } catch (error) {
      console.error("Failed to get hot suburbs:", error);
      throw new Error("Failed to retrieve high growth suburbs");
    }
  }

  /**
   * Get market predictions
   */
  async getMarketPredictions(): Promise<unknown> {
    try {
      const response = await apiService.get(
        `${this.baseUrl}/market/predictions`,
      );
      return response.data;
    } catch (error) {
      console.error("Failed to get market predictions:", error);
      throw new Error("Failed to retrieve market predictions");
    }
  }

  /**
   * Get market risk factors
   */
  async getMarketRiskFactors(): Promise<unknown> {
    try {
      const response = await apiService.get(
        `${this.baseUrl}/market/risk-factors`,
      );
      return response.data;
    } catch (error) {
      console.error("Failed to get market risk factors:", error);
      throw new Error("Failed to retrieve market risk factors");
    }
  }
}

// Export singleton instance
export const propertyIntelligenceService = PropertyIntelligenceService
  .getInstance();

File: backend/app/router/property_intelligence.py
"""
Property Intelligence API Router for Real2.AI
Advanced property analysis, market intelligence, and investment insights
"""

from typing import List, Optional, Dict, Any
from fastapi import (
    APIRouter,
    HTTPException,
    status,
    Depends,
    Query,
    Path,
    BackgroundTasks,
)
from fastapi.responses import StreamingResponse
from pydantic import BaseModel, Field, field_validator
import logging
from datetime import datetime, timedelta
import asyncio
import json
from io import StringIO
import csv

from app.schema import (
    PropertySearchRequest,
    PropertyProfile,
)

logger = logging.getLogger(__name__)


# Additional schema models for property intelligence
class PropertyAnalyticsRequest(BaseModel):
    """Request model for property analytics"""

    properties: List[str] = Field(..., description="List of property addresses or IDs")
    analysis_type: str = Field("comprehensive", description="Type of analysis")
    include_market_comparison: bool = True
    include_investment_metrics: bool = True
    include_risk_assessment: bool = True

    @field_validator("analysis_type")
    @classmethod
    def validate_analysis_type(cls, v):
        valid_types = ["basic", "standard", "comprehensive"]
        if v not in valid_types:
            raise ValueError(f"Analysis type must be one of: {valid_types}")
        return v


class PropertyInvestmentRecommendation(BaseModel):
    """Property investment recommendation"""

    recommendation_type: str
    confidence_score: float
    reasoning: List[str]
    key_factors: List[str]
    risk_warnings: List[str]
    optimal_holding_period: str
    expected_return_range: Dict[str, float]


class PropertyMarketInsight(BaseModel):
    """Property market insight"""

    insight_id: str
    insight_type: str
    title: str
    description: str
    impact_level: str
    affected_areas: List[str]
    time_horizon: str
    confidence_level: str
    data_sources: List[str]
    created_at: datetime
    expires_at: Optional[datetime] = None


class PropertyAnalyticsResponse(BaseModel):
    """Property analytics response"""

    request_id: str
    properties_analyzed: int
    analysis_type: str
    property_profiles: List[PropertyProfile]
    recommendations: List[PropertyInvestmentRecommendation]
    market_insights: List[PropertyMarketInsight]
    data_quality_score: float
    processing_time: float
    total_cost: float
    created_at: datetime


class BulkPropertyAnalysisRequest(BaseModel):
    """Bulk property analysis request"""

    properties: List[str] = Field(..., min_length=2, max_length=50)
    analysis_depth: str = Field("standard", description="Analysis depth")
    output_format: str = Field("json", description="Output format")

    @field_validator("analysis_depth")
    @classmethod
    def validate_analysis_depth(cls, v):
        valid_depths = ["basic", "standard", "comprehensive"]
        if v not in valid_depths:
            raise ValueError(f"Analysis depth must be one of: {valid_depths}")
        return v


class PropertyWatchlistRequest(BaseModel):
    """Property watchlist request"""

    address: str = Field(..., description="Property address")
    notes: Optional[str] = None
    tags: List[str] = []
    alert_preferences: Dict[str, bool] = {
        "price_changes": True,
        "market_updates": False,
        "similar_listings": False,
    }


class PropertyWatchlistResponse(BaseModel):
    """Property watchlist response"""

    id: str
    property: Optional[PropertyProfile] = None
    saved_at: datetime
    notes: Optional[str] = None
    tags: List[str] = []
    alert_preferences: Dict[str, bool]
    is_favorite: bool = False
    price_alerts_triggered: int = 0
    last_price_change: Optional[datetime] = None


class PropertyComparisonResult(BaseModel):
    """Property comparison result"""

    comparison_id: str
    properties: List[PropertyProfile]
    comparison_matrix: Dict[str, Dict[str, Any]]
    rankings: Dict[str, List[str]]
    summary_insights: List[str]
    recommendation: str
    created_at: datetime


router = APIRouter(prefix="/property-intelligence", tags=["Property Intelligence"])


# Dependency injection for services
async def get_property_service():
    """Get property intelligence service"""
    from app.services.property.property_intelligence_service import PropertyIntelligenceService

    return PropertyIntelligenceService()


async def get_market_service():
    """Get market intelligence service"""
    from app.services.property.market_intelligence_service import MarketIntelligenceService

    return MarketIntelligenceService()


async def get_current_user():
    """Get current authenticated user"""
    # Placeholder for actual user authentication
    return {"id": "user_123", "subscription_tier": "premium"}


# API Endpoints


@router.post(
    "/search",
    response_model=Dict[str, Any],
    summary="Advanced Property Search",
    description="""
    Search properties with advanced filters and intelligent matching.
    
    Features:
    - Natural language query processing
    - Geographic radius search
    - Advanced property filters
    - Intelligent ranking and relevance scoring
    - Real-time market data integration
    
    **Rate Limits:**
    - Free tier: 10 searches/hour
    - Premium tier: 100 searches/hour
    """,
)
async def search_properties(
    request: PropertySearchRequest,
    background_tasks: BackgroundTasks,
    service=Depends(get_property_service),
    current_user=Depends(get_current_user),
):
    """Advanced property search with intelligent filtering"""

    try:
        # Log search for analytics
        background_tasks.add_task(
            log_property_search,
            user_id=current_user["id"],
            search_params=request.dict(),
        )

        # Mock response for development
        search_results = {
            "search_id": f"search_{datetime.now().strftime('%Y%m%d_%H%M%S')}",
            "query": request.query,
            "total_results": 47,
            "results_returned": min(request.limit, 47),
            "search_time_ms": 234,
            "properties": [],
            "facets": {
                "price_ranges": {
                    "0-500000": 12,
                    "500000-750000": 18,
                    "750000-1000000": 11,
                    "1000000+": 6,
                },
                "property_types": {"House": 28, "Unit": 12, "Townhouse": 7},
                "bedrooms": {"1": 3, "2": 14, "3": 20, "4+": 10},
            },
            "market_summary": {
                "median_price": 785000,
                "price_trend": "increasing",
                "market_activity": "active",
            },
        }

        # Generate mock property results
        for i in range(min(request.limit, 47)):
            property_data = {
                "id": f"prop_{i+1}",
                "address": f"{123 + i} Example Street, Parramatta NSW 2150",
                "price": 650000 + (i * 25000),
                "bedrooms": 2 + (i % 3),
                "bathrooms": 1 + (i % 2),
                "carspaces": i % 3,
                "property_type": ["House", "Unit", "Townhouse"][i % 3],
                "land_area": 450 + (i * 10) if i % 3 == 0 else None,
                "building_area": 120 + (i * 5),
                "market_score": 85 - (i * 2),
                "investment_score": 78 + (i % 10),
                "listing_date": (datetime.now() - timedelta(days=i * 3)).isoformat(),
                "estimated_rental": 520 + (i * 15),
            }
            search_results["properties"].append(property_data)

        return search_results

    except Exception as e:
        logger.error(f"Property search failed: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Property search failed",
        )


@router.post(
    "/analyze",
    response_model=PropertyAnalyticsResponse,
    summary="Comprehensive Property Analysis",
    description="""
    Generate comprehensive property analysis with market intelligence.
    
    Analysis includes:
    - Professional property valuation
    - Investment metrics and ROI calculations
    - Risk assessment and factors
    - Market trends and forecasting
    - Neighborhood analysis
    - Comparable sales analysis
    - Financial modeling scenarios
    
    **Costs:**
    - Basic analysis: $8-12 AUD
    - Standard analysis: $15-25 AUD
    - Comprehensive analysis: $25-40 AUD
    """,
)
async def analyze_property(
    request: PropertyAnalyticsRequest,
    background_tasks: BackgroundTasks,
    service=Depends(get_property_service),
    current_user=Depends(get_current_user),
):
    """Comprehensive property analysis"""

    try:
        # Validate request
        if not request.properties:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="At least one property is required for analysis",
            )

        # Start background analysis tracking
        analysis_id = f"analysis_{datetime.now().strftime('%Y%m%d_%H%M%S')}"

        background_tasks.add_task(
            track_property_analysis,
            analysis_id=analysis_id,
            user_id=current_user["id"],
            properties=request.properties,
            analysis_type=request.analysis_type,
        )

        # Mock comprehensive analysis response
        analysis_response = PropertyAnalyticsResponse(
            request_id=analysis_id,
            properties_analyzed=len(request.properties),
            analysis_type=request.analysis_type,
            property_profiles=[],  # Would be populated by actual service
            recommendations=[
                PropertyInvestmentRecommendation(
                    recommendation_type="buy",
                    confidence_score=85.5,
                    reasoning=[
                        "Strong capital growth potential in area",
                        "Below-market purchase price opportunity",
                        "High rental demand demographics",
                    ],
                    key_factors=[
                        "Transport infrastructure development",
                        "Population growth trend",
                        "Limited housing supply",
                    ],
                    risk_warnings=[
                        "Interest rate sensitivity",
                        "Market cycle considerations",
                    ],
                    optimal_holding_period="3-7 years",
                    expected_return_range={"min": 6.5, "max": 12.8},
                )
            ],
            market_insights=[
                PropertyMarketInsight(
                    insight_id="insight_001",
                    insight_type="trend",
                    title="Strong Growth Trajectory",
                    description="Market showing sustained growth with infrastructure investment driving demand",
                    impact_level="high",
                    affected_areas=["Parramatta", "Westmead", "Harris Park"],
                    time_horizon="medium_term",
                    confidence_level="high",
                    data_sources=["domain", "corelogic", "abs"],
                    created_at=datetime.now(),
                )
            ],
            data_quality_score=0.92,
            processing_time=3.45,
            total_cost=28.50,
            created_at=datetime.now(),
        )

        return analysis_response

    except Exception as e:
        logger.error(f"Property analysis failed: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Property analysis failed",
        )


@router.post(
    "/bulk-analyze",
    response_model=Dict[str, Any],
    summary="Bulk Property Portfolio Analysis",
    description="""
    Analyze multiple properties for portfolio management and comparison.
    
    Features:
    - Portfolio-level metrics and insights
    - Diversification analysis
    - Risk correlation analysis
    - Performance benchmarking
    - Bulk export capabilities
    
    **Limits:**
    - Maximum 50 properties per request
    - Premium subscription required
    """,
)
async def bulk_analyze_properties(
    request: BulkPropertyAnalysisRequest,
    background_tasks: BackgroundTasks,
    service=Depends(get_property_service),
    current_user=Depends(get_current_user),
):
    """Bulk property analysis for portfolio management"""

    try:
        # Check subscription tier
        if current_user["subscription_tier"] not in ["premium", "enterprise"]:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="Bulk analysis requires premium subscription",
            )

        # Start bulk analysis
        batch_id = f"bulk_{datetime.now().strftime('%Y%m%d_%H%M%S')}"

        background_tasks.add_task(
            process_bulk_analysis,
            batch_id=batch_id,
            user_id=current_user["id"],
            properties=request.properties,
            analysis_depth=request.analysis_depth,
        )

        # Return processing confirmation
        return {
            "batch_id": batch_id,
            "properties_queued": len(request.properties),
            "estimated_completion_minutes": len(request.properties) * 2,
            "status": "processing",
            "analysis_depth": request.analysis_depth,
            "estimated_cost": len(request.properties) * 15.0,
            "result_formats": (
                ["json", "csv", "pdf"]
                if request.output_format == "json"
                else [request.output_format]
            ),
        }

    except Exception as e:
        logger.error(f"Bulk analysis failed: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Bulk property analysis failed",
        )


@router.get(
    "/watchlist",
    response_model=List[PropertyWatchlistResponse],
    summary="Get User Property Watchlist",
    description="Retrieve user's saved properties and watchlist with alerts",
)
async def get_property_watchlist(
    limit: int = Query(50, ge=1, le=100),
    offset: int = Query(0, ge=0),
    current_user=Depends(get_current_user),
):
    """Get user's property watchlist"""

    try:
        # Mock watchlist data
        watchlist_items = []

        for i in range(min(limit, 10)):
            item = PropertyWatchlistResponse(
                id=f"watch_{i+1}",
                property=None,  # Would be populated with actual PropertyProfile
                saved_at=datetime.now() - timedelta(days=i * 7),
                notes=(
                    f"Interested in this property for investment purposes"
                    if i % 2 == 0
                    else None
                ),
                tags=["investment", "growth-area"] if i % 3 == 0 else ["family-home"],
                alert_preferences={
                    "price_changes": True,
                    "market_updates": True,
                    "similar_listings": False,
                },
                is_favorite=i < 3,
                price_alerts_triggered=i % 4,
                last_price_change=(
                    datetime.now() - timedelta(days=i * 2) if i % 3 == 0 else None
                ),
            )
            watchlist_items.append(item)

        return watchlist_items

    except Exception as e:
        logger.error(f"Failed to retrieve watchlist: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to retrieve property watchlist",
        )


@router.post(
    "/watchlist",
    response_model=Dict[str, Any],
    summary="Add Property to Watchlist",
    description="Save property to user's watchlist with custom alerts",
)
async def add_to_watchlist(
    request: PropertyWatchlistRequest, current_user=Depends(get_current_user)
):
    """Add property to user's watchlist"""

    try:
        # Mock watchlist addition
        watchlist_item = {
            "id": f"watch_{datetime.now().strftime('%Y%m%d_%H%M%S')}",
            "address": request.address,
            "saved_at": datetime.now().isoformat(),
            "notes": request.notes,
            "tags": request.tags,
            "alert_preferences": request.alert_preferences,
            "status": "saved",
        }

        return watchlist_item

    except Exception as e:
        logger.error(f"Failed to add property to watchlist: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to add property to watchlist",
        )


@router.get(
    "/market-insights",
    response_model=List[PropertyMarketInsight],
    summary="Get Market Insights",
    description="Retrieve market insights and trends for specified locations",
)
async def get_market_insights(
    location: str = Query(..., description="Location (suburb, city, or postcode)"),
    insight_types: List[str] = Query(
        ["trends", "forecasts"], description="Types of insights to retrieve"
    ),
    limit: int = Query(10, ge=1, le=50),
    market_service=Depends(get_market_service),
):
    """Get market insights for location"""

    try:
        # Mock market insights
        insights = []

        insight_templates = [
            {
                "type": "trend",
                "title": "Strong Price Growth Momentum",
                "description": f"{location} showing 8.5% price growth over 12 months with increasing buyer activity",
                "impact": "high",
            },
            {
                "type": "forecast",
                "title": "Continued Growth Expected",
                "description": f"Market forecasting 5-7% growth for {location} over next 12 months",
                "impact": "medium",
            },
            {
                "type": "opportunity",
                "title": "Infrastructure Investment Impact",
                "description": f"Upcoming transport projects expected to boost {location} property values",
                "impact": "high",
            },
        ]

        for i, template in enumerate(insight_templates[:limit]):
            if template["type"] in insight_types:
                insight = PropertyMarketInsight(
                    insight_id=f"insight_{location}_{i+1}",
                    insight_type=template["type"],
                    title=template["title"],
                    description=template["description"],
                    impact_level=template["impact"],
                    affected_areas=[location],
                    time_horizon="medium_term",
                    confidence_level="high",
                    data_sources=["domain", "corelogic", "abs"],
                    created_at=datetime.now(),
                    expires_at=datetime.now() + timedelta(days=30),
                )
                insights.append(insight)

        return insights

    except Exception as e:
        logger.error(f"Failed to retrieve market insights: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to retrieve market insights",
        )


@router.post(
    "/compare",
    response_model=PropertyComparisonResult,
    summary="Compare Properties",
    description="Side-by-side comparison of multiple properties with analysis",
)
async def compare_properties(
    properties: List[str] = Query(
        ..., min_length=2, max_length=10, description="Property addresses or IDs"
    ),
    comparison_criteria: List[str] = Query(
        ["price", "investment", "growth"], description="Comparison criteria"
    ),
    service=Depends(get_property_service),
):
    """Compare multiple properties side by side"""

    try:
        # Mock property comparison
        comparison_id = f"comp_{datetime.now().strftime('%Y%m%d_%H%M%S')}"

        comparison_result = PropertyComparisonResult(
            comparison_id=comparison_id,
            properties=[],  # Would be populated with actual PropertyProfile objects
            comparison_matrix={
                "price": {
                    prop: 650000 + i * 50000 for i, prop in enumerate(properties)
                },
                "investment_score": {
                    prop: 75 + i * 5 for i, prop in enumerate(properties)
                },
                "growth_potential": {
                    prop: 85 - i * 3 for i, prop in enumerate(properties)
                },
            },
            rankings={
                "price": properties.copy(),  # sorted by price
                "investment": properties[::-1],  # reverse for investment score
                "overall": properties.copy(),
            },
            summary_insights=[
                f"Property 1 offers best value for money at current market prices",
                f"Property 2 shows strongest investment fundamentals",
                f"All properties in comparison show above-average growth potential",
            ],
            recommendation=f"Based on analysis, {properties[0]} offers optimal risk-adjusted returns",
            created_at=datetime.now(),
        )

        return comparison_result

    except Exception as e:
        logger.error(f"Property comparison failed: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Property comparison failed",
        )


@router.get(
    "/export/{format}",
    summary="Export Property Data",
    description="Export property analysis data in various formats",
)
async def export_property_data(
    format: str = Path(..., pattern="^(csv|json|pdf)$"),
    properties: List[str] = Query(..., description="Property IDs to export"),
    current_user=Depends(get_current_user),
):
    """Export property data in requested format"""

    try:
        if format == "csv":
            # Generate CSV data
            output = StringIO()
            writer = csv.writer(output)

            # Headers
            writer.writerow(
                [
                    "Address",
                    "Price",
                    "Bedrooms",
                    "Bathrooms",
                    "Investment Score",
                    "Risk Level",
                ]
            )

            # Mock data
            for i, prop_id in enumerate(properties):
                writer.writerow(
                    [
                        f"123 Example St, Suburb NSW 2000",
                        650000 + i * 50000,
                        2 + i % 3,
                        1 + i % 2,
                        85 - i * 2,
                        ["Low", "Medium", "High"][i % 3],
                    ]
                )

            csv_data = output.getvalue()
            output.close()

            return StreamingResponse(
                iter([csv_data]),
                media_type="text/csv",
                headers={
                    "Content-Disposition": "attachment; filename=property_data.csv"
                },
            )

        elif format == "json":
            # Generate JSON data
            data = {
                "export_date": datetime.now().isoformat(),
                "properties": [
                    {
                        "id": prop_id,
                        "address": f"123 Example St, Suburb NSW 2000",
                        "price": 650000 + i * 50000,
                        "analysis": {"investment_score": 85 - i * 2},
                    }
                    for i, prop_id in enumerate(properties)
                ],
            }

            return StreamingResponse(
                iter([json.dumps(data, indent=2)]),
                media_type="application/json",
                headers={
                    "Content-Disposition": "attachment; filename=property_data.json"
                },
            )

        else:  # PDF
            raise HTTPException(
                status_code=status.HTTP_501_NOT_IMPLEMENTED,
                detail="PDF export not yet implemented",
            )

    except Exception as e:
        logger.error(f"Export failed: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Export operation failed",
        )


# Background task functions
async def log_property_search(user_id: str, search_params: Dict[str, Any]):
    """Log property search for analytics"""
    logger.info(f"Property search by user {user_id}: {search_params}")


async def track_property_analysis(
    analysis_id: str, user_id: str, properties: List[str], analysis_type: str
):
    """Track property analysis for billing and analytics"""
    logger.info(
        f"Property analysis {analysis_id} by user {user_id}: {len(properties)} properties, type: {analysis_type}"
    )


async def process_bulk_analysis(
    batch_id: str, user_id: str, properties: List[str], analysis_depth: str
):
    """Process bulk property analysis in background"""
    logger.info(
        f"Bulk analysis {batch_id} by user {user_id}: {len(properties)} properties, depth: {analysis_depth}"
    )
    # Simulate processing
    await asyncio.sleep(2)
    logger.info(f"Bulk analysis {batch_id} completed")
